### PR TITLE
Windows

### PR DESCRIPTION
**Pull request by bam241**
This should allow to build dagmc on windows.

Still a draft as this needs a new version of PyNE. (right now pulling the one from my fork.)

### CODE REVIEW

**Comment by bam241** [comment]
@gonuke @pshriwise this can probably be reviewed while we wait for the  PR pyne/pyne#1367 to be merged.

File: cmake/DAGMC_macros.cmake
Comment (gonuke) [inline cmake/DAGMC_macros.cmake]: Hmmm... isn't CMake supposed to get this kind of thing right? I wonder why we set it explicitly in the first place?

File: cmake/FindMOAB.cmake
Comment (gonuke) [inline cmake/FindMOAB.cmake:43]: Do shared libs fail, or are they just unnecessary?

File: src/CMakeLists.txt
Comment (gonuke) [inline src/CMakeLists.txt:11]: Do we only need PyNE for UWUW?

File: src/make_watertight/MakeWatertight.cpp
Comment (gonuke) [inline src/make_watertight/MakeWatertight.cpp]: Can't you `resize()` this before the loop?

File: src/pyne/update_pyne.sh
Comment (gonuke) [inline src/pyne/update_pyne.sh]: This needs to be updated

File: src/uwuw/uwuw.cpp
Comment (gonuke) [inline src/uwuw/uwuw.cpp]: Still interested in getting uniform behavior here...

**Review by gonuke — CHANGES_REQUESTED** [review]
Thanks @bam241 - I have a few questions/comments.

File: cmake/FindMOAB.cmake
Comment (bam241) [reply cmake/FindMOAB.cmake:43]: unnecessary if I recall correctly

File: src/CMakeLists.txt
Comment (bam241) [reply src/CMakeLists.txt:11]: I believe so yes, so we might as well group them together

**Comment by gonuke** [comment]
Looks like this needs a rebase, esp. wrt PyNE

File: src/uwuw/uwuw.cpp
Comment (bam241) [reply src/uwuw/uwuw.cpp]: I still don't know what to do for this one....

Comment (gonuke) [reply src/uwuw/uwuw.cpp]: let's leave it for now

Comment (gonuke) [reply src/uwuw/uwuw.cpp]: Although the simple option is to have all platforms behave like the Windows one.  It is more limiting, but consistent

**Comment by gonuke** [comment]
This looks like it's getting closer.  I wonder if the problem now is that we use forward slash '/' in the file names in `amalgamate_pyne.sh`?

**Comment by bam241** [comment]
I believe the problem occurs when trying to run the bash scripts (on load, before it actually try to run what is in it...)

I found some hint of answer Friday, didn't had time to try them yesterday....

File: src/uwuw/uwuw.cpp
Comment (gonuke) [reply src/uwuw/uwuw.cpp]: Other than the problems caused by this in PyNE, wouldn't `filesystem::canonical()` work here?

Comment (bam241) [reply src/uwuw/uwuw.cpp]: yes I think so
was planing on fixing this if we had a fix for the PyNE compilation

**Comment by bam241** [comment]
I have PR both in MOAB and PyNE to fixe this.

https://bitbucket.org/fathomteam/moab/pull-requests/550/fixing-cpp17-compatibilities 
and pyne/pyne#1390

File: .github/workflows/windows.yml
Comment (bam241) [inline .github/workflows/windows.yml]: this has to be cleaned before merge !

File: .github/workflows/windows.yml
Comment (bam241) [inline .github/workflows/windows.yml]: this need to be clean as well

File: .gitmodules
Comment (bam241) [inline .gitmodules]: this should be reverted when PyNE PR gets merged pyne/pyne#1390

File: .tmp_circle/config.yml
Comment (bam241) [inline .tmp_circle/config.yml]: this file should be removed.

File: cmake/DAGMC_macros.cmake
Comment (bam241) [inline cmake/DAGMC_macros.cmake]: this should be removed again....

**Review by bam241 — COMMENTED** [review]
I can't request change (as it is my PR) but all those comments must be addressed before merge

**Comment by bam241** [comment]
I think we are having a problem:
```
bash /root/build_dir/DAGMC/src/pyne/amalgamate_pyne.sh /root/build_dir/DAGMC/src/pyne /root/build_dir/DAGMC/src/pyne
Traceback (most recent call last):
  File "amalgamate.py", line 145, in <module>
    main()
  File "amalgamate.py", line 125, in main
    hdr.append_file(f)
  File "amalgamate.py", line 75, in append_file
    with open(filename, 'rt', encoding='utf-8') as f:
TypeError: 'encoding' is an invalid keyword argument for this function
```

**Comment by gonuke** [comment]
I relaunched these tests to see if it helps... but it may need a rebase

**Review by gonuke — APPROVED** [review]

**Comment by bam241** [comment]
we probably need to wait for  the MOAB 5.3.0 release

File: .github/workflows/windows.yml
Comment (bam241) [inline .github/workflows/windows.yml]: this needs to change to moab 5.3.0

**Comment by bam241** [comment]
@gonuke I set MOAB version to 5.3.0.

It is good on my side, let me know what you still think that's good to go  :)

File: .github/workflows/windows.yml
Comment (gonuke) [inline .github/workflows/windows.yml]: Do we want any path guards here?

File: .github/workflows/windows.yml
Comment (gonuke) [inline .github/workflows/windows.yml]: Is this the newest version of HDF5 that works?

File: .github/workflows/windows.yml
Comment (gonuke) [inline .github/workflows/windows.yml]: Can we conda install HDF5 (since we're in a conda environment anyway?)

File: .github/workflows/windows.yml
Comment (gonuke) [inline .github/workflows/windows.yml]: can we run tests?

File: src/pyne/remove_unsupported.py
Comment (gonuke) [inline src/pyne/remove_unsupported.py]: is this because windows is stuck with Py 2?

**Review by gonuke — CHANGES_REQUESTED** [review]
A few new comments with fresh eyes and new experiences.

File: src/pyne/remove_unsupported.py
Comment (bam241) [reply src/pyne/remove_unsupported.py]: I don't know, but with the new change I PRed this shouldn't be necessary....

File: .github/workflows/windows.yml
Comment (bam241) [reply .github/workflows/windows.yml]: I don't know.
I think I unsuccessfully tried.
let me try again :)

File: src/pyne/remove_unsupported.py
Comment (gonuke) [reply src/pyne/remove_unsupported.py]: should we try without? or is this necessary for some users stuck on Py2 systems?

**Comment by bam241** [comment]
@gonuke you were correct conda hdf5 works !

File: .github/workflows/windows.yml
Comment (bam241) [reply .github/workflows/windows.yml]: I don't know how to run the test.

make is not available, so I don't know how to run them...

**Comment by bam241** [comment]
I am not sure how to run the test... I think we build them tho.


I'll try using ctest...

File: .github/workflows/windows.yml
Comment (gonuke) [inline .github/workflows/windows.yml]: Need to turn this on?

**Review by gonuke — CHANGES_REQUESTED** [review]
Maybe we need to turn on tests in CMake?

File: .github/workflows/windows.yml
Comment (bam241) [reply .github/workflows/windows.yml]: that would make sense :)

**Comment by gonuke** [comment]
You may need to wget the test files like we do on linux

**Comment by bam241** [comment]
I don't think we download them... at least not all of them...
`ls` shows them in the `bin`/`test` folder:
```
check_watertight.exe
cones.h5m
cyl.h5m
dag_file.h5m
dagmc_pointinvol_test.exe
dagmc_rayfire_test.exe
dagmc_simple_test.exe
dagmc_unit_tests.exe
make_watertight.exe
make_watertight_cone_tests.exe
make_watertight_cylinder_tests.exe
make_watertight_no_curve_sphere_tests.exe
make_watertight_sphere_n_box_test.exe
mat_lib.h5
MOAB.dll
no_curve_sphere.h5m
pt_vol_test.exe
ray_fire_test.exe
sphere_n_box.h5m
test_dagmc.h5m
test_dagmc_impl.h5m
test_geom.exe
test_geom.h5m
test_uwuw.h5m
uwuw_preproc.exe
uwuw_unit_tests.exe
uwuw_unit_tests_preprocessor.exe
uwuw_unit_tests_tally.exe
```

**Comment by gonuke** [comment]
The Linux install script has a wget line that the windows stuff doesn't include

**Comment by bam241** [comment]
@gonuke I finally got that working :)

**Comment by gonuke** [comment]
Is it necessary to include all of gtest in the PR? or can we install it from source somehow?

File: cmake/DAGMC_macros.cmake
Comment (gonuke) [inline cmake/DAGMC_macros.cmake]: Can't we use C++17 on Linux?

**Comment by bam241** [comment]
it is not new, it was like that before, I just need from `1.7.0` to `1.8.0`

File: cmake/DAGMC_macros.cmake
Comment (gonuke) [inline cmake/DAGMC_macros.cmake:236]: Cleanup white space?

File: cmake/test_config/CMakeLists.txt
Comment (gonuke) [inline cmake/test_config/CMakeLists.txt]: See above comment

**Comment by gonuke** [comment]
I just realized that this is just an upgrade!

File: .github/workflows/windows.yml
Comment (gonuke) [inline .github/workflows/windows.yml]: Does this copy in all the MOAB libs & includes? If so, are we faking the tests? Should we be able to build against MOAB where it is installed?

**Review by gonuke — CHANGES_REQUESTED** [review]
It's great that we got a working version... now to refine it.

File: .github/workflows/windows.yml
Comment (bam241) [reply .github/workflows/windows.yml]: There is multiple thing there, but you have point:
- I took what I had for the plugin which build every library separately and then combines them into the plugin
- we need a default install folder (like `~/.local` on linux) and add it to the `windows` path

basically I don't think we are faking the test but sloppy/dirty :) 
I'll work on cleaning this up

**Comment by gonuke** [comment]
Also - can we upgrade gtest as a separate PR, first to confine these changes to the Windows stuff?

**Comment by gonuke** [comment]
> Also - can we upgrade gtest as a separate PR, first to confine these changes to the Windows stuff?

And do it as a delete & add rather than a mv

**Comment by gonuke** [comment]
Rebase for #778

**Comment by gonuke** [comment]
How did this break? 😢

Was it moving the gtest into a separate PR?

**Comment by bam241** [comment]
No basically I can get the right way to had the test into the PATH

B. Mouginot

**Comment by gonuke** [comment]
This latest update didn't seem to launch actions from your `windows` branch

**Comment by gonuke** [comment]
Huh - I think it's because you don't launch on push for this workflow.

**Comment by bam241** [comment]
@gonuke I fixed it !

File: .github/workflows/windows.yml
Comment (gonuke) [inline .github/workflows/windows.yml]: Isn't this the same?

```suggestion
          echo "PATH=$PATH:/d/a/DAGMC/install_dir/bin:/d/a/DAGMC/install_dir/tests" >> $GITHUB_ENV
```

**Review by gonuke — CHANGES_REQUESTED** [review]
Great to finally figure this out!  Just one little question about that path...

File: .github/workflows/windows.yml
Comment (bam241) [reply .github/workflows/windows.yml]: yes ... 
image

**Comment by gonuke** [comment]
One last cleanup request.... rename the tasks in a consistent way:
- "Standard Build and Test / BuildTest" -> "Linux Build and Test / BuildTest"
- "Windows / main" -> "Windows Build and Test / BuildTest"

File: .github/workflows/windows_build_test.yml
Comment (gonuke) [inline .github/workflows/windows_build_test.yml]:
```suggestion
  BuildTest:
```

**Review by gonuke — CHANGES_REQUESTED** [review]
One last name change

**Comment by gonuke** [comment]
I'm not sure whether or not this needs a rebase?

**Comment by bam241** [comment]
I don't think so, there is no conflict.
there is 1 overlapping change, but they were the exact same... (a missing `#include <array>` ... )

**Comment by gonuke** [comment]
This is such a great step forward! Thanks for tracking this all down @bam241

**Pull request merged**
